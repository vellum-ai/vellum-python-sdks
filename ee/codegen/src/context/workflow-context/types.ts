export interface SDK_MODULE_PATHS {
  WORKFLOWS_MODULE_PATH: readonly string[];
  CORE_NODES_MODULE_PATH: readonly string[];
  DISPLAYABLE_NODES_MODULE_PATH: readonly string[];
  INPUTS_MODULE_PATH: readonly string[];
  SANDBOX_RUNNER_MODULE_PATH: readonly string[];
  STATE_MODULE_PATH: readonly string[];
  NODE_DISPLAY_MODULE_PATH: readonly string[];
  NODE_DISPLAY_TYPES_MODULE_PATH: readonly string[];
  VELLUM_TYPES_MODULE_PATH: readonly string[];
  PORTS_MODULE_PATH: readonly string[];
}
