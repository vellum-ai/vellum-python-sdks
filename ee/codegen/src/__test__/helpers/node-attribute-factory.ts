import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
