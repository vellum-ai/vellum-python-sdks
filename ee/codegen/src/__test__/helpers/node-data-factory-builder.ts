import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
