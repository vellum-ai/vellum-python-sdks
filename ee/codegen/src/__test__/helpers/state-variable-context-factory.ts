import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
