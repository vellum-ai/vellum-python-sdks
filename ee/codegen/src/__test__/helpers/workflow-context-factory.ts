import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
