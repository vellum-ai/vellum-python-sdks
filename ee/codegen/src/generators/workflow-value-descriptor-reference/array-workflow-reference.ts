import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";

import { BaseNodeInputWorkflowReference } from "./BaseNodeInputWorkflowReference";

import { WorkflowValueDescriptor } from "src/generators/workflow-value-descriptor";
import { ArrayWorkflowReference as ArrayWorkflowReferenceType } from "src/types/vellum";

export class ArrayWorkflowReference extends BaseNodeInputWorkflowReference<ArrayWorkflowReferenceType> {
  getAstNode(): AstNode | undefined {
    const arrayReference = this.nodeInputWorkflowReferencePointer;
    const items = arrayReference.items;

    if (!items || items.length === 0) {
      return python.TypeInstantiation.list([]);
    }

    const listItems = items.map((item) => {
      const valueNode = new WorkflowValueDescriptor({
        nodeContext: this.nodeContext,
        workflowContext: this.workflowContext,
        workflowValueDescriptor: item,
        iterableConfig: this.iterableConfig,
        attributeConfig: this.attributeConfig,
      });

      return valueNode;
    });

    return python.TypeInstantiation.list(listItems, {
      endWithComma: true,
    });
  }
}
