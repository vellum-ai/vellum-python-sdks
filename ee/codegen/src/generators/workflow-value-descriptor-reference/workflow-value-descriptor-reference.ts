import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";

import { WorkflowContext } from "src/context";
import { ValueGenerationError } from "src/generators/errors";
import { BaseNodeInputWorkflowReference } from "src/generators/workflow-value-descriptor-reference/BaseNodeInputWorkflowReference";
import { ConstantValueReference } from "src/generators/workflow-value-descriptor-reference/constant-value-reference";
import { ExecutionCounterWorkflowReference } from "src/generators/workflow-value-descriptor-reference/execution-counter-workflow-reference";
import { NodeOutputWorkflowReference } from "src/generators/workflow-value-descriptor-reference/node-output-workflow-reference";
import { VellumSecretWorkflowReference } from "src/generators/workflow-value-descriptor-reference/vellum-secret-workflow-reference";
import { WorkflowInputReference } from "src/generators/workflow-value-descriptor-reference/workflow-input-reference";
import {
  IterableConfig,
  WorkflowValueDescriptorReference as WorkflowValueDescriptorReferenceType,
} from "src/types/vellum";
import { assertUnreachable } from "src/utils/typing";

export declare namespace WorkflowValueDescriptorReference {
  export interface Args {
    workflowContext: WorkflowContext;
    workflowValueReferencePointer: WorkflowValueDescriptorReferenceType;
    iterableConfig?: IterableConfig;
  }
}

export class WorkflowValueDescriptorReference extends AstNode {
  private workflowContext: WorkflowContext;
  public readonly workflowValueReferencePointer: WorkflowValueDescriptorReferenceType["type"];
  private iterableConfig?: IterableConfig;
  public astNode:
    | BaseNodeInputWorkflowReference<WorkflowValueDescriptorReferenceType>
    | undefined;

  constructor(args: WorkflowValueDescriptorReference.Args) {
    super();

    this.workflowContext = args.workflowContext;
    this.workflowValueReferencePointer =
      args.workflowValueReferencePointer.type;
    this.iterableConfig = args.iterableConfig;

    this.astNode = this.getAstNode(args.workflowValueReferencePointer);

    if (this.astNode) {
      this.inheritReferences(this.astNode);
    }
  }

  private getAstNode(
    workflowValueReferencePointer: WorkflowValueDescriptorReferenceType
  ):
    | BaseNodeInputWorkflowReference<WorkflowValueDescriptorReferenceType>
    | undefined {
    const referenceType = workflowValueReferencePointer.type;

    switch (referenceType) {
      case "NODE_OUTPUT": {
        const reference = new NodeOutputWorkflowReference({
          workflowContext: this.workflowContext,
          nodeInputWorkflowReferencePointer: workflowValueReferencePointer,
        });
        if (reference.getAstNode()) {
          return reference;
        } else {
          return undefined;
        }
      }
      case "WORKFLOW_INPUT":
        return new WorkflowInputReference({
          workflowContext: this.workflowContext,
          nodeInputWorkflowReferencePointer: workflowValueReferencePointer,
        });
      case "WORKFLOW_STATE":
        this.workflowContext.addError(
          new ValueGenerationError(
            "WORKFLOW_STATE reference pointers is not implemented"
          )
        );
        return undefined;
      case "CONSTANT_VALUE":
        return new ConstantValueReference({
          workflowContext: this.workflowContext,
          nodeInputWorkflowReferencePointer: workflowValueReferencePointer,
          iterableConfig: this.iterableConfig,
        });
      case "VELLUM_SECRET":
        return new VellumSecretWorkflowReference({
          workflowContext: this.workflowContext,
          nodeInputWorkflowReferencePointer: workflowValueReferencePointer,
        });
      case "EXECUTION_COUNTER":
        return new ExecutionCounterWorkflowReference({
          workflowContext: this.workflowContext,
          nodeInputWorkflowReferencePointer: workflowValueReferencePointer,
        });
      default: {
        assertUnreachable(referenceType);
      }
    }
  }

  public write(writer: Writer): void {
    if (this.astNode) {
      this.astNode.write(writer);
    }
  }
}
