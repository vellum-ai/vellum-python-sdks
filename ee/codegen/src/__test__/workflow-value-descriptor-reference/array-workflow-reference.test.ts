import { Writer } from "@fern-api/python-ast/core/Writer";
import { vi } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { WorkflowContext } from "src/context";
import { ArrayWorkflowReference } from "src/generators/workflow-value-descriptor-reference/array-workflow-reference";
import { ArrayWorkflowReference as ArrayWorkflowReferenceType } from "src/types/vellum";

describe("ArrayWorkflowReference", () => {
  let workflowContext: WorkflowContext;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("should generate correct AST for empty array", async () => {
    const emptyArrayReference: ArrayWorkflowReferenceType = {
      type: "ARRAY_REFERENCE",
      items: [],
    };

    const reference = new ArrayWorkflowReference({
      workflowContext,
      nodeInputWorkflowReferencePointer: emptyArrayReference,
    });

    const writer = new Writer();
    reference.write(writer);

    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for array with constant values", async () => {
    const arrayReference: ArrayWorkflowReferenceType = {
      type: "ARRAY_REFERENCE",
      items: [
        {
          type: "CONSTANT_VALUE",
          value: {
            type: "STRING",
            value: "Hello, World!",
          },
        },
        {
          type: "CONSTANT_VALUE",
          value: {
            type: "NUMBER",
            value: 42,
          },
        },
      ],
    };

    const reference = new ArrayWorkflowReference({
      workflowContext,
      nodeInputWorkflowReferencePointer: arrayReference,
    });

    const writer = new Writer();
    reference.write(writer);

    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for nested arrays", async () => {
    const nestedArrayReference: ArrayWorkflowReferenceType = {
      type: "ARRAY_REFERENCE",
      items: [
        {
          type: "ARRAY_REFERENCE",
          items: [
            {
              type: "CONSTANT_VALUE",
              value: {
                type: "STRING",
                value: "Inner value",
              },
            },
            {
              type: "CONSTANT_VALUE",
              value: {
                type: "NUMBER",
                value: 100,
              },
            },
          ],
        },
        {
          type: "CONSTANT_VALUE",
          value: {
            type: "STRING",
            value: "Outer value",
          },
        },
      ],
    };

    const reference = new ArrayWorkflowReference({
      workflowContext,
      nodeInputWorkflowReferencePointer: nestedArrayReference,
    });

    const writer = new Writer();
    reference.write(writer);

    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should generate correct AST for array with expression values", async () => {
    const exprArrayReference: ArrayWorkflowReferenceType = {
      type: "ARRAY_REFERENCE",
      items: [
        {
          type: "UNARY_EXPRESSION",
          operator: "null",
          lhs: {
            type: "CONSTANT_VALUE",
            value: {
              type: "STRING",
              value: "check if null",
            },
          },
        },
        {
          type: "BINARY_EXPRESSION",
          operator: "=",
          lhs: {
            type: "CONSTANT_VALUE",
            value: {
              type: "NUMBER",
              value: 10,
            },
          },
          rhs: {
            type: "CONSTANT_VALUE",
            value: {
              type: "NUMBER",
              value: 10,
            },
          },
        },
      ],
    };

    const reference = new ArrayWorkflowReference({
      workflowContext,
      nodeInputWorkflowReferencePointer: exprArrayReference,
    });

    const writer = new Writer();
    reference.write(writer);

    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });
});
