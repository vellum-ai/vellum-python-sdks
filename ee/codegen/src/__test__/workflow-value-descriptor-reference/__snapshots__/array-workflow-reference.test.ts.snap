// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArrayWorkflowReference > should generate correct AST for array with constant values 1`] = `
"[
    "Hello, World!",
    42,
]
"
`;

exports[`ArrayWorkflowReference > should generate correct AST for array with expression values 1`] = `
"[
    ConstantValueReference("check if null").is_null(),
    ConstantValueReference(10).equals(10),
]
"
`;

exports[`ArrayWorkflowReference > should generate correct AST for empty array 1`] = `
"[]
"
`;

exports[`ArrayWorkflowReference > should generate correct AST for nested arrays 1`] = `
"[
    [
        "Inner value",
        100,
    ],
    "Outer value",
]
"
`;
